### ISSUE TITLE
CsrfAuthenticationStrategy with stateless double submit CSRF token not working as intended?

### ISSUE THREAD

**Issue by high-stakes** [spring-projects/spring-security#5669]
### Summary
Hi,

CsrfAuthenticationStrategy only seems to support stateful csrf setups as it creates a new token on each authentication as in stateless setups onauthenticate is called for each request. Considering the recommended CSRF token repository for Angular apps (which usually use stateless token based authentication) is the CookieCsrfTokenRepository which uses the stateless double submit cookie validation pattern, this looks like erroneous behavior as in stateless mode I would not expect to have a new CSRF token generated every request.

### Actual Behavior

When using CookieCsrfTokenRepository in a stateless setting the CSRF token is recreated on each request because onAuthenticate is always triggered due to no active session found

### Expected Behavior

because the token is stateless it is only recreated when no CSRF cookie is sent in the response when there is no http session. The token should not be tied to authentication.

### Configuration

```
http
	.authorizeRequests()
		.anyRequest().authenticated()
	.and()
	.sessionManagement()
	       .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
	.and()
	.addFilterBefore(new CustomAuthenticationTokenFilter(), AnonymousAuthenticationFilter.class);

```
### Version

Spring boot 2.0.3

**Comment by rwinch** [comment]
@high-stakes Thanks for the feedback.

`CsrfAuthenticationStrategy` only applies to after a user successfully authenticates. This is intended to avoid attacks similar to a session fixation attack where a malicious user establishes a valid CSRF token and then tricks a user into authenticating and then using the CSRF token from prior to authentication.

I'm not sure I understand your use case. Can you explain what you are trying to do in terms of HTTP request/responses?

**Comment by high-stakes** [comment]
Hi @rwinch ,

My question is basically in the case of a sessionless application which uses an authentication token that is verified on each request and triggers the onAuthentication method on each request causing a new CSRF token to be generated always even though the user is already "authenticated" (not in a stateful sense though just from the client pov).

Now with the double submit CSRF token solution which is stateless the only thing that is verified if the request header matches the request csrf cookie. The potential of token fixation does not decrease in this case by only sending it when authenticated because you don't check the session contents to verify the csrf token.

I was just wondering if this is intended as it seems accidental to tie a stateless verification to the authentication state.

**Comment by rwinch** [comment]
It is quite deliberate that the CSRF token changes on authentication success. This is necessary.

Can you expand on what you mean by stateless? If you are using authentication that is not included by a browser automatically (i.e. an OAuth token), then you do not need to enable CSRF protection. However, if you are using something like HTTP Basic, there is still a risk.

**Comment by lburja** [comment]
For reference, if others run into this, you can use `httpSecurity.csrf().sessionAuthenticationStrategy(new NullAuthenticatedSessionStrategy()) ` to work around this issue.

**Comment by jalik** [comment]
@lburja Dude, thank you so much for your help. I was trying to make Spring work with a SPA based on React, in which case we obviously don't want to have a different CSRF token generated on every single AJAX request... a thing that @rwinch seemed to not understand being blocked on token rotation during auth, while @high-stakes was talking about requests post auth.

**Comment by OtenMoten** [comment]
The thread is closed but the behaviour is the same. The workaround of @lburja works! Is there any official solution?

**Comment by maxmmin** [comment]
@lburja thank for your answer, i love you man. You made my day and mood so much better. Thank you man❤️

**Comment by claudiumaravela** [comment]
@OtenMoten you’re completely right, but if you’re generating that token in your backend and you store it as a httpOnly secure cookie, then you’re vulnerable to csrf, right? Because that secure cookie is attached to all the requests (by the browser).

**Comment by OtenMoten** [comment]
In a previous comment, I mentioned that _"In a stateless Spring Boot security application with OAuth 2 bearer authorization via JWT, you do not need to implement Cross-Site Request Forgery (CSRF) protection."_ While this is true for scenarios where JWT tokens are passed in the Authorization header and not stored in cookies, it does not hold if the tokens are stored as httpOnly secure cookies.

If you store your JWT token in an httpOnly secure cookie, you are vulnerable to CSRF attacks because the browser attaches the secure cookie to all requests made to the same domain. This means an attacker could craft a malicious request to exploit the user's authenticated session.

To mitigate CSRF attacks in this scenario, you should implement CSRF protection mechanisms. One common approach is the double-submit cookie pattern, where a CSRF token is sent both as a cookie and as a request header. The server then verifies that the values match.

If you generate a token in your backend and store it as an httpOnly secure cookie, you are vulnerable to CSRF attacks. This vulnerability arises because the secure cookie is automatically included in all requests by the browser, allowing potential attackers to exploit it. Therefore, you need to implement CSRF protection to safeguard against such attacks.

**Comment by OtenMoten** [comment]
> @OtenMoten you’re completely right, but if you’re generating that token in your backend and you store it as a httpOnly secure cookie, then you’re vulnerable to csrf, right? Because that secure cookie is attached to all the requests (by the browser).

[claudiumaravela](https://github.com/claudiumaravela)

In a previous comment, I mentioned that _"In a stateless Spring Boot security application with OAuth 2 bearer authorization via JWT, you do not need to implement Cross-Site Request Forgery (CSRF) protection."_ While this is true for scenarios where JWT tokens are passed in the Authorization header and not stored in cookies, it does not hold if the tokens are stored as httpOnly secure cookies.

If you store your JWT token in an httpOnly secure cookie, you are vulnerable to CSRF attacks because the browser attaches the secure cookie to all requests made to the same domain. This means an attacker could craft a malicious request to exploit the user's authenticated session.

To mitigate CSRF attacks in this scenario, you should implement CSRF protection mechanisms. One common approach is the double-submit cookie pattern, where a CSRF token is sent both as a cookie and as a request header. The server then verifies that the values match.

If you generate a token in your backend and store it as an httpOnly secure cookie, you are vulnerable to CSRF attacks. This vulnerability arises because the secure cookie is automatically included in all requests by the browser, allowing potential attackers to exploit it. Therefore, you need to implement CSRF protection to safeguard against such attacks.
